compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/TypeReference.kt

```diff
@@ -158,21 +158,32 @@ public sealed class TypeReference {
      * interface ServiceFactory : Factory<Service>
      * ```
      *
-     * The KtTypeReference `T` will fail to resolve, since it isn't a type. This function will
+     * The type reference `T` will fail to resolve, since it isn't a type. This function will
      * instead look to the `ServiceFactory` interface, then look at the supertype declaration
      * in order to determine the type.
      *
+     * It's possible that this method may still return a generic type, if `T` resolves to another
+     * generic type. For instance, `T` will resolve to `R` in the following example:
+     *
+     * ```
+     * interface Factory<T>
+     *
+     * interface UnknownFactory<R> : Factory<R>
+     * ```
+     *
      * @param implementingClass The class which actually references the type. In the above
      * example, this would be `ServiceFactory`.
      */
     internal fun resolveTypeReference(
       implementingClass: ClassReference
-    ): TypeReference? {
+    ): TypeReference {
       // If the element isn't a type variable name like `T`, it can be resolved through imports.
       type.typeElement?.fqNameOrNull(module)
         ?.let { return this }
 
-      return resolveGenericTypeReference(implementingClass, declaringClass, type.text)
+      // When we fail to resolve the generic type here, we're potentially at the end of a chain of
+      // generics, so we return the current type instead of null.
+      return resolveGenericTypeReference(implementingClass, declaringClass, type.text) ?: this
```

Comment:
With this change, a number of `orNull` style methods in `TypeReference` can no longer return null. I wanted to see how you feel about this change before cleaning those up, Ralf. As far as I can tell, we don't currently rely on being able to get null returned when the generic can't be resolved. The cases that were using the `orNull` calls appear to then try and get the class reference which will end up null anyway.

Reply:
Actually looks like a numer of the `orNull`s are still valid, so I only udpated the Psi one where it's always non-null now.